README.md
## 介绍
- 如果你的项目中需要很多的表格页，但没有太多复杂的业务逻辑，那么你可以试一下 react-fast-table
- 传入四个接口就生成一个简单的表格页面，减少前端的压力或可以将这种类似的表格页，完全交由后端配置
- 接口目前默认你使用正向代理跨域，所以接口不需要带服务器地址和端口
  - getDataList，获取表格数据
  - insertItem，新增
  - removeItem，删除
  - updateItem，修改
- 如果接口需要判断身份，则需要自己配置 headers，配置的 headers 将作为请求头
## 下载
`yarn add react-fast-table`
## 使用
- 前端代码
```javascript
import React from 'react'
import FastTable from 'react-fast-table'

const App = () => (
  <FastTable
    getDataList='/getList'
    insertItem='/insertItem'
    removeItem='/removeItem'
    updateItem='/updateItem'
    headers={{
      'Accept': 'application/json',
      'Content-Type': 'application/json',
      'cache-control': 'no-store'
    }}
  />
)

export default App
```
- 接口数据格式
  - getDataList
    - list: 表格数据
    ```json
    [
        {
            "id": {
                "value": 0,
                "isShow": false
            },
            "name": {
                "value": "test0",
                "isShow": true
            },
            "age": {
                "value": 20,
                "isShow": true
            },
            "sex": {
                "value": 1,
                "isShow": true
            },
            "date": {
                "value": "2020-01-01",
                "isShow": true
            },
            "date2": {
                "value": [
                    "2020-01-01",
                    "2020-01-02"
                ],
                "isShow": true
            }
        }
    ]
    ```
    - pagination: 分页，具体配置可参照 antd 的 pagination
    ```json
    "pagination": {
        "pageSize": 20,
        "total": 10,
        "showQuickJumper": true,
        "showSizeChanger": true
    }
    ```
    - header: table 表头和表单元素的配置，字段名要与 list 的字段名对应，设置为 false 时不显示
      - title： 显示在表头的文字
      - inputOptions：需要展示的表单，用于查询，新增，修改
        - type：表单元素类型
          - Input：普通输入框
          - InputNumber：数字输入框
          - Select：下拉框，type 为 Select 需要为其配置 enums，详情见下方
          - DatePicker：日期控件，数据格式：'2020-01-01'
          - RangePicker：时间范围控件，数据格式：['2020-01-01', '2020-01-02'],
        - required：是否为必填项，'true | false'
        - defaultValue：表单元素默认值
        - enums：当 type 为 Select 时，配置此为其枚举选项
        - placeholder：提示内容
        - isSearch： 是否需要作为搜索选项
    ```json
    "header": {
        "id": false,
        "name": {
            "title": "姓名",
            "inputOptions": {
                "type": "Input",
                "required": false,
                "defaultValue": "",
                "enums": [],
                "placeholder": "请输入姓名",
                "isSearch": true
            }
        },
        "age": {
            "title": "年龄",
            "inputOptions": {
                "type": "InputNumber",
                "required": true,
                "defaultValue": null,
                "placeholder": "请输入年龄",
                "isSearch": true
            }
        },
        "sex": {
            "title": "性别",
            "inputOptions": {
                "type": "Select",
                "required": true,
                "defaultValue": 1,
                "placeholder": "请输入性别",
                "isSearch": true,
                "enums": {
                    "0": "女",
                    "1": "男"
                }
            }
        },
        "date": {
            "title": "时间",
            "inputOptions": {
                "type": "DatePicker",
                "required": true,
                "defaultValue": "2001",
                "isSearch": true
            }
        },
        "date2": {
            "title": "两个时间",
            "inputOptions": {
                "type": "RangePicker",
                "required": true,
                "defaultValue": [
                    "2001",
                    "2002"
                ],
                "isSearch": true
            }
        }
    }
    ```
getDataList 数据实例
```javascript
{
    "list": [
        {
            "id": {
                "value": 0,
                "isShow": false
            },
            "name": {
                "value": "test0",
                "isShow": true
            },
            "age": {
                "value": 20,
                "isShow": true
            },
            "sex": {
                "value": 1,
                "isShow": true
            },
            "date": {
                "value": "2020-01-01",
                "isShow": true
            },
            "date2": {
                "value": [
                    "2020-01-01",
                    "2020-01-02"
                ],
                "isShow": true
            }
        }
    ],
    "header": {
        "id": false,
        "name": {
            "title": "姓名",
            "inputOptions": {
                "type": "Input",
                "required": false,
                "defaultValue": "",
                "enums": [],
                "placeholder": "请输入姓名",
                "isSearch": true
            }
        },
        "age": {
            "title": "年龄",
            "inputOptions": {
                "type": "InputNumber",
                "required": true,
                "defaultValue": null,
                "placeholder": "请输入年龄",
                "isSearch": true
            }
        },
        "sex": {
            "title": "性别",
            "inputOptions": {
                "type": "Select",
                "required": true,
                "defaultValue": 1,
                "placeholder": "请输入性别",
                "isSearch": true,
                "enums": {
                    "0": "女",
                    "1": "男"
                }
            }
        },
        "date": {
            "title": "时间",
            "inputOptions": {
                "type": "DatePicker",
                "required": true,
                "defaultValue": "2001",
                "isSearch": true
            }
        },
        "date2": {
            "title": "两个时间",
            "inputOptions": {
                "type": "RangePicker",
                "required": true,
                "defaultValue": [
                    "2001",
                    "2002"
                ],
                "isSearch": true
            }
        }
    },
    "pagination": {
        "pageSize": 20,
        "total": 10,
        "showQuickJumper": true,
        "showSizeChanger": true
    }
}
```
  - insertItem
    - 请求参数
    ```json
    {
      "input": {
        "name": "1234",
        "age": 12,
        "sex": 1,
        "date": "2001-01-01 12:00:00",
        "date2": [
          "2001-01-01 12:00:00",
          "2002-01-01 12:00:00"
        ]
      }
    }
    ```
    - 返回参数
    ```json
    {"status": "success"}
    ```
  - updateItem，与 insertItem 基本一致
  - removeItem
    - 请求参数
    ```json
    {"input": 0}
    ```
    - 返回参数
    ```json
    {status: "success"}
    ```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.8](https://github.com/SZZH/react-fast-table/compare/v1.0.7...v1.0.8) (2020-09-04)


src/index.js
/*
 * @Author: zenghao
 * @Date: 2020-07-28 13:57:04
 * @LastEditTime: 2020-09-04 21:24:00
 * @LastEditors: zenghao
 * @Description: 
 * @FilePath: /ReactFastTable/src/index.js
 * @Copyright 2020 OBKoro1
 */

import React, { useState, useEffect } from 'react'
import { Button, Spin, Input, InputNumber, Select, DatePicker, ConfigProvider } from 'antd'
import moment from 'moment'
import 'moment/locale/zh-cn'
import locale from 'antd/es/locale/zh_CN'

import TableBody from './TableBody'
import EditModal from './EditModal'
import SearchForm from './SearchForm'
import { setHeaders } from './request'
import { getDataList, insertItem, removeItem, updateItem, setUri } from './service'
import { changeDataType } from './kits'

import './index.css'
const { RangePicker } = DatePicker

const formatMomentObj = obj => {
  const formatTemple = 'YYYY-MM-DD'
  let tempObj
  if(typeof obj !== 'object') return obj
  if(Array.isArray(obj)) {
    tempObj = [...obj]
    for(let i in tempObj) {
      if(typeof tempObj[i] === 'object' && !tempObj[i].hasOwnProperty('_isAMomentObject')) return obj
      tempObj[i] = tempObj[i].format(formatTemple)
    }
  }else {
    if(!obj.hasOwnProperty('_isAMomentObject')) return obj
    tempObj = obj.format(formatTemple)
  }

  return tempObj
}

const timeInputValue = {
  DatePicker: defaultValue => moment(defaultValue),
  RangePicker: defaultValue => [moment(defaultValue[0]), moment(defaultValue[1])]
}

// 获取编辑页面的 input 配置
const getAllInput = (header) => {
  let allInput = []

  for (let key in header) {
    // header 中无 key 或没有配置 inputOptions 时不添加 input
    if (!header[key] || !header[key].inputOptions) continue
    
    const type = header[key].inputOptions.type
    // 日期类型的 input ，需要经过 moment 格式化
    const defaultValue = timeInputValue.hasOwnProperty(type) &&
      timeInputValue[type](header[key].inputOptions.defaultValue) ||
      header[key].inputOptions.defaultValue

    allInput.push({
      ...header[key].inputOptions,
      key,
      title: header[key].title,
      type: header[key].inputOptions.type,
      defaultValue,
      required: header[key].inputOptions.required,
      enums: header[key].inputOptions.enums,
    })
  }
  return allInput
}

// 生成表单元素
const inputEnums = {
  'Input': options => <Input {...options}  style={{width: '100%'}}/>,
  'InputNumber': options => <InputNumber {...options}  style={{width: '100%'}}/>,
  'Select': options => (
    <Select {...options} style={{width: '100%'}}>
      {
        (() => {
          let resultArr = []
          for(let key in options.enums) {
            resultArr.push(
              <Select.Option
                value={changeDataType(options.defaultValue, key)}
              >
                {options.enums[key]}
              </Select.Option>
            )
          }
          return resultArr
        })()
      }
    </Select>
  ),
  'DatePicker': options => <DatePicker {...options}/>,
  'RangePicker': options => <RangePicker {...options}/>
}

// 页面配置初始化
const initPage = props => {
  setUri(props)
  setHeaders(props.headers)
}

const formData = {}

const FastTable = props => {
  initPage(props)
  const [isShowEditModal, setIsShowEditModal] = useState(false)
  const [editModalType, setEditModalType] = useState('add')
  const [data, setData] = useState({
    list: [],
    header: {},
    pagintion: {}
  })
  const [loading, setLoading] = useState(false)
  const [editSubmitLoading, setEditSubmitLoading] = useState(false)
  const [editData, setEditData] = useState({})

  // 保存 searchFormValue
  const setFormData = newFormData => {
    for(let key in newFormData) {
      formData[key] = newFormData[key]
    }
  }

  // 更新 data
  const updateData = async (body = { current: 1, pageSize: 20, input: {} }) => {
    setFormData(body.input)
    setLoading(true)
    const { list, pagination, header } = await getDataList({...body, input: formData})
    setData({
      list,
      header,
      pagination
    })
    setLoading(false)
  }

  // 编辑 modal 的取消事件
  const handleCancelEditModal = () => {
    setIsShowEditModal(false)
  }

  // 编辑的提交事件
  const handleEditSubmit = async formValues => {
    // 格式化数据，若为时间则格式化为 YYYY-MM-DD hh:mm:ss
    let tempFormValues = {...formValues}
    for(let key in tempFormValues) {
      tempFormValues[key] = formatMomentObj(tempFormValues[key])
    }
    
    setEditSubmitLoading(true)
    if (editModalType === 'add') {
      await insertItem({ input: tempFormValues })
    } else {
      await updateItem({
        input: { id: editData.id, ...tempFormValues }
      })
    }
    setEditSubmitLoading(false)
    updateData()
    handleCancelEditModal()
  }

  // 删除事件
  const handleRemoveItem = async ({ current, pageSize, id }) => {
    await removeItem({ input: id })
    updateData({ current, pageSize })
  }

  // 打开编辑 modal
  const handleShowEditModal = (type, record) => {
    setIsShowEditModal(true)
    setEditModalType(type)
    if (!record) return
    setEditData(record)
  }

  // 编辑 model 关闭的回调
  const handleModalOnClosed = () => {
    setEditData({})
  }

  useEffect(() => {
    updateData()
  }, [])

  return (
    <ConfigProvider locale={locale}>
      <Spin spinning={loading}>
        <div
          style={{
            overflow: 'auto',
            margin: '5px 0',
            padding: 10,
            width: '100%',
          }}
        >
          <Button
            type='primary'
            onClick={() => handleShowEditModal('add')}
          >
            新增
          </Button>
          <SearchForm
            allInput={getAllInput(data.header)}
            inputEnums={inputEnums}
            updateData={updateData}
            setFormData={setFormData}
          />
          <TableBody
            dataSource={data.list}
            pagination={data.pagination}
            tableHeader={data.header}
            removeItem={handleRemoveItem}
            updateData={updateData}
            handleShowEditModal={handleShowEditModal}
            style={{
              margin: '5px 0'
            }}
          />
          <EditModal
            title={editModalType === 'add' && '新增' || '编辑'}
            editModalType={editModalType}
            afterClose={handleModalOnClosed}
            allInput={getAllInput(data.header)}
            visible={isShowEditModal}
            handleEditSubmit={handleEditSubmit}
            onCancel={handleCancelEditModal}
            updateData={updateData}
            editSubmitLoading={editSubmitLoading}
            editData={editData}
            inputEnums={inputEnums}
          />
        </div>
      </Spin>
    </ConfigProvider>
  )
}

export default FastTable